src/derivers/cip3Icarus.ts

```diff
@@ -0,0 +1,348 @@
+ *
+ * - HD node consists of a 64 byte private key, 32 byte public key and 32 byte chain code.
+ * - See https://github.com/cardano-foundation/CIPs/blob/09d7d8ee1bd64f7e6b20b5a6cae088039dce00cb/CIP-0003/CIP-0003.md#master-key-generation.
+ *
+ * - For derivation of BIP32 HD nodes, it uses modified version called BIP32-Ed25519.
+ * - See https://input-output-hk.github.io/adrestia/static/Ed25519_BIP.pdf.
+ */
+
+// Native BigInt uses big-endian. Since cip3Icarus(bip32Edd25519) uses little-endian, we need to reverse the bytes
+// and have separate functions for bigIntToBytes and bytesToBigInt, .slice() is used just to make a copy of the array
+const toReversed = (bytes: Uint8Array) => bytes.slice().reverse();
+
+const bytesToBigInt = (bytes: Uint8Array) => {
+  const reversed = toReversed(bytes);
+  const bytesInHex = bytesToHex(reversed);
+  return BigInt(bytesInHex);
+};
+
+const bigIntToBytes = (bigInt: bigint) => {
+  const hexadecimal = bigInt.toString(16);
+  return toReversed(hexToBytes(hexadecimal));
+};
+
+const padEnd32Bytes = (bytes: Uint8Array) => {
+  return concatBytes([
+    bytes,
+    new Uint8Array(Math.max(32 - bytes.length, 0)).fill(0),
+  ]);
+};
+
+/**
+ * Truncates to first 28 bytes and multiplies by 8.
+ *
+ * @param bytes - Little-Endian big number in bytes.
+ * @returns PadEnd32Bytes(left[0, 28] * 8)).
+ */
+const trunc28Mul8 = (bytes: Uint8Array): Uint8Array => {
+  const truncLeftMul8 = bytesToBigInt(bytes.slice(0, 28)) * BigInt(8);
+  return padEnd32Bytes(bigIntToBytes(truncLeftMul8));
+};
+
+/**
+ * Does module 2^256.
+ *
+ * @param bytes - Little-Endian big number in bytes.
+ * @returns PadEnd32Bytes(mod(bytes, 2^256))).
+ */
+const mod2Pow256 = (bytes: Uint8Array): Uint8Array => {
+  return padEnd32Bytes(
+    bigIntToBytes(mod(bytesToBigInt(bytes), BigInt(2) ** BigInt(256))),
+  );
+};
+
+/**
+ * Adds the left to the right.
+ *
+ * @param left - Left hand side Little-Endian big number.
+ * @param right - Right hand side Little-Endian big number.
+ * @returns PadEnd32Bytes(left + right).
+ */
+const add = (left: Uint8Array, right: Uint8Array): Uint8Array => {
+  const added = bytesToBigInt(left) + bytesToBigInt(right);
+  return padEnd32Bytes(bigIntToBytes(added)).slice(0, 32);
+};
+
+/**
+ * Concat tag, key and childIndex.
+ *
+ * @param tag - Key specific tag.
+ * @param key - Key.
+ * @param childIndex - Child index.
+ * @returns PadEnd32Bytes(left + right).
+ */
+const getKeyExtension = (tag: number, key: Uint8Array, childIndex: number) => {
+  return concatBytes([
+    new Uint8Array([tag]),
+    key,
+    numberToUint32(childIndex, true),
+  ]);
+};
+
+type DeriveKeyBaseArgs = { childIndex: number };
+
+type DeriveWithPrivateArgs = DeriveKeyBaseArgs & {
+  parentNode: {
+    privateKeyBytes: Uint8Array;
+    chainCodeBytes: Uint8Array;
+    publicKeyBytes: Uint8Array;
+  };
+  isHardened: boolean;
+};
+
+const Z_TAGS = {
+  normal: 2,
```

Comment:
"normal" is imho arguable here, I'd go with `nonHardened` and `hardened`, I don't think there's any consensus on which of these derivations is normal, they are both just fit for different usecases